tensorflow/core/kernels/data/cache_dataset_ops.cc

```diff
@@ -275,6 +288,9 @@ class CacheDatasetOp::FileDataset : public DatasetBase {
 
       Status SaveInternal(IteratorStateWriter* writer) override {
         mutex_lock l(mu_);
+        TF_RETURN_IF_ERROR(
```

Comment:
The old implementation makes the following case break:
1. Finish the iteration of `FileWriterIterator` and save it using `SaveInternal()`
2. When restoring `FileIterator`, `FileIterator::RestoreInternal()` will create a `FileReaderIterator` instead of `FileWriterIterator`, but it is still restored based on the serialized data at the first step. 
3. Currently, the saving order in `FileWriterIterator::SaveInternal()` does not match with `FileReaderIterator::RestoreInternal()`.

This change makes `cur_index_ ` be saved before other elements in order to be compatible with [FileReaderIterator:: RestoreInternal()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/cache_dataset_ops.cc#L507).

Reply:
Great find! Can we add a test case to prevent this bug from happening again?

Reply:
Yeah, the enhanced Roundtrip test ([here](https://github.com/tensorflow/tensorflow/blob/80223ea8a01718df61891fc8a23645fc02829edc/tensorflow/core/kernels/data/cache_dataset_ops_test.cc#L463)) has covered this case now by adding [more breakpoints](https://github.com/tensorflow/tensorflow/pull/30825/files#diff-7def5e1766c22b54ae33c7479b478a9cR105). When `breakpoint` < `cardinality`, the restored iterator will be `FileWriterIterator`; When `breakpoint` > `cardinality`,  the restored iterator will be `FileReaderIterator `.

Reply:
got it, thanks